### PR TITLE
[ENH] Expand grid refactor (WIP)

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

-  Reuse existing functions for type checking
-

**This PR improves #617 .**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [x] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [x] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [x] Ensure that you have added tests.
- [x] Run all tests (`$ pytest .`) locally on your machine.
    - [x] Check to ensure that test coverage covers the lines of code that you have added.
    - [x] Ensure that all tests pass.

## Documentation Changes

If you are adding documentation changes, please ensure the following:

- [x] Build the docs locally.
- [x] View the docs to check that it renders correctly.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Review by ericmjl — APPROVED** [review]

**Comment by ericmjl** [comment]
Approving ahead-of-time, @samukweku! I noticed one or two places where the docstrings aren't adhering to semantic line breaks, will go in and mark them. Otherwise, it's well-documented, looks well-tested, all tests pass, and test coverage goes up, so it passes the bar for me at the moment :smile:.

File: janitor/utils.py [hunk 1]
```diff
@@ -410,133 +410,104 @@ def _wrapped(x, *args, **kwargs):
...
-    How each type is handled, and their associated exceptions,
-    are pretty clear from the code.
+    This function relies on itertools' `product` function
+    to create the different combinations. It is possible
+    that in the future, a faster implementation will be used,
+    and still respect data types.
+    In this function, a couple of checks are conducted as well.
+    This include type checking, as well as the size (if it is a
+    numpy array, it should not be greater than 2). Dataframes
+    with MultiIndex columns are not supported.
```
Comment (ericmjl) [inline janitor/utils.py]: As mentioned in the overall comment, marking out this paragraph to be changed to SEMBR-compliant form.

File: janitor/functions.py [hunk 2]
```diff
@@ -4041,15 +4040,16 @@ def expand_grid(
...
-    a key to represent the column name in the output must be provided.
+    This works with a dictionary of name value pairs, and will work
+    with structures that are not dataframes.
 
-    Note that if a MultiIndex dataframe or series is passed, the index/columns
+    If method-chaining to a dataframe, a key to represent the column
+    name in the output must be provided.
+
+    Note that if a MultiIndex dataframe or series is passed, the index
     will be discarded, and a single indexed dataframe will be returned.
```
Comment (ericmjl) [inline janitor/functions.py]: As mentioned in the overall comment, marking out this paragraph to be changed to SEMBR-compliant form.

**Review by ericmjl — APPROVED** [review]

**Comment by samukweku** [comment]
@ericmjl made changes based on the SEMBR suggestion. Thanks! not sure I have it figured out completely though.

**Comment by samukweku** [comment]
@ericmjl and the rest of the team, if it's ok with you, can we leave this PR unmerged for a while. Hopefully I can improve on the speed

**Comment by ericmjl** [comment]
Got it, @samukweku! I'm going to add "WIP" to the title so that I don't accidentally merge the thing :smile:.

**Comment by samukweku** [comment]
closing this; starting afresh.

**Pull request closed without merging**